File: strings/string-to-integer/string-to-integer-atoi.java
class Solution {
  public int myAtoi(String s) {
    char[] chars = s.toCharArray();

    int i = 0;
    int sign = 1;
    while (i < chars.length && chars[i] == ' ') {
      i++;
    }

    if (i < chars.length && chars[i] == '-') {
      sign = -1;
      i++;
    } else if (i < chars.length && chars[i] == '+') {
      i++;
    }

    long sum = 0;

    while (i < chars.length && chars[i] >= '0' && chars[i] <= '9') {
      sum = (sum * 10) + (chars[i] - '0') * sign;
      if (sum > Integer.MAX_VALUE) {
        return Integer.MAX_VALUE;
      } else if (sum < Integer.MIN_VALUE) {
        return Integer.MIN_VALUE;
      }
      i++;
    }

    return (int) sum;

  }
}

File: strings/string-to-integer/string-to-integer-atoi.ts
function myAtoi(s: string): number {
  const chars: string[] = s.split('');

  let i: number = 0;
  let sign: number = 1;

  while (i < chars.length && chars[i] === ' ') {
    i++;
  }

  if (i < chars.length && chars[i] === '-') {
    sign = -1;
    i++;
  } else if (i < chars.length && chars[i] === '+') {
    i++;
  }

  let sum = 0;

  while (i < chars.length && chars[i] >= '0' && chars[i] <= '9') {
    sum = sum * 10 + (chars[i].charCodeAt(0) - '0'.charCodeAt(0)) * sign;

    if (sum > 2 ** 31 - 1) {
      return 2 ** 31 - 1;
    }
    if (sum < -(2 ** 31)) {
      return -(2 ** 31);
    }
    i++;
  }

  return sum;
}
